### DISCUSSION TITLE
Exported stitched_ip project is depended on /tmp/ files.

### DISCUSSION THREAD

**Discussion by laurenslejeune** [Xilinx/finn#404] [Q&A] [answered]
Hello,
I've been running into a puzzling issue when outputting any of my projects to _stitched_ip_. When using such a stitched_ip project in Vivado, I am able to produce the following result in behavioral testing:
![pre_restart](https://user-images.githubusercontent.com/44036367/137978025-f7da698e-39c9-48c0-8ba1-71369ca6b5a9.png)
This result is reproducible, regardless of whether FINN or Vivado are restarted. However, as soon as I restart my system, the exact same experiment no longer returns data:
![post_restart](https://user-images.githubusercontent.com/44036367/137978298-27aa5e37-1486-4539-b183-f01f0407e066.png)
Even though the _m_valid_ still outputs correctly, the _m_data_-signal remains XXX (or unknown). This bug occurs for any of my projects, regardless of the model or specific FINN transformations.

I figure that, to the best of my knowledge, the only thing that specifically changes when the system restarts is the /tmp/ folder. And indeed, if I copy the files of my /tmp/finn_dev_laurens/ directory to another location after building my _stitched_ip_, and copy the same directory back to /tmp/ after system restart, I am able to counteract the issue. As this fix is not very practical, and as this does not seem like something that should be necessary, I wonder what I can do to fix the issue.

System: Ubuntu 20.04.3 LTS (Focal Fossa)
Vivado version: 2020.1
FINN version: v0.5b
I realize I am not using the latest version of FINN, but as everything I need to work right now works (and does not work out-of-the-box for v0.6), I hope that is not an issue. I am planning to upgrade to v0.6 soon, though. Thanks in advance for any advice.

**Comment by fpjentzsch** [comment, marked as answer]
Hi,
you can set the environment variable `FINN_HOST_BUILD_DIR` to define a location outside of /tmp.
For example: `export FINN_HOST_BUILD_DIR=/home/laurens/tmp_finn`

FINN places all intermediate results/projects in this directory, so you should probably clean it up once in a while if you run many builds.

**Reply by laurenslejeune** [reply]
Hi,
I can confirm this also solves the issue. However, does this mean that I should retain the generated files for each project that I want to be able to keep using? Should I then try to manually remove intermediate results?

**Reply by fpjentzsch** [reply]
Yes, you should retain those to work/experiment/debug "inside" the corresponding build. Normally, the build flow only exports a few output products, such as bitstream, driver, and intermediate ONNX models. The intermediate models contain absolute paths to the underlying tmp directories for code generation, Vivado, or HLS projects (e.g. for every node in your model).

You don't need to remove anything in the FINN_HOST_BUILD_DIR. It will just grow in size as it fills with the files of multiple builds.

**Comment by maltanar** [comment]
Another option (if you use the latest `dev` branch or the soon-to-come v0.7 release) is to use the newly-introduced `stitched_ip_gen_dcp` option as part of the `build_dataflow` config, this will generate a synthesized .dcp checkpoint as part of the stitched-IP export and should no longer need the individual layer files:

https://github.com/Xilinx/finn/blob/dev/src/finn/builder/build_dataflow_config.py#L206

**Reply by laurenslejeune** [reply]
Hi Yaman, I will check out that build then. Thanks for the information!
